stateproof/builder.go

```diff
@@ -20,48 +20,163 @@ import (
+		if proto.StateProofInterval == 0 || uint64(rnd)%proto.StateProofInterval != 0 {
+			continue
+		}
+
+		builderExists, err := spw.builderExists(rnd)
+		if err != nil {
+			spw.log.Warnf("OnPrepareVoterCommit(%d): could not check builder existence, assuming it doesn't exist: %v\n", rnd, err)
+		} else if builderExists {
+			continue
+		}
+
+		buildr, err := createBuilder(rnd, votersFetcher)
+		if err != nil {
+			if errors.Is(err, errVotersNotTracked) {
+				// Voters not tracked for that round.  Might not be a valid
+				// state proof round; state proofs might not be enabled; etc.
+				spw.log.Warnf("OnPrepareVoterCommit(%d): %v", rnd, err)
+				continue
+			}
+
+			spw.log.Errorf("OnPrepareVoterCommit(%d): could not create builder: %v", rnd, err)
+			continue
+		}
+
+		err = spw.db.Atomic(func(_ context.Context, tx *sql.Tx) error {
+			return persistBuilder(tx, rnd, &buildr)
+		})
+		if err != nil {
+			spw.log.Errorf("OnPrepareVoterCommit(%d): could not persist builder: %v", rnd, err)
+		}
+	}
+}
+
+// loadOrCreateBuilderWithSignatures either loads a builder from the DB or creates a new builder.
+// this function fills the builder with all the available signatures
+func (spw *Worker) loadOrCreateBuilderWithSignatures(rnd basics.Round) (builder, error) {
+	b, err := spw.loadOrCreateBuilder(rnd)
+	if err != nil {
+		return builder{}, err
+	}
+
+	if err := spw.loadSignaturesIntoBuilder(&b); err != nil {
+		return builder{}, err
+	}
+	return b, nil
+}
+
+func (spw *Worker) loadOrCreateBuilder(rnd basics.Round) (builder, error) {
+	var buildr builder
+	err := spw.db.Atomic(func(ctx context.Context, tx *sql.Tx) (err error) {
+		buildr, err = getBuilder(tx, rnd)
+		return err
+	})
+
+	if err == nil {
+		return buildr, nil
+	}
+
+	if !errors.Is(err, sql.ErrNoRows) {
+		spw.log.Errorf("loadOrCreateBuilder: error while fetching builder from DB: %v", err)
+	}
+
+	buildr, err = createBuilder(rnd, spw.ledger)
+	if err != nil {
+		return builder{}, err
+	}
+
+	err = spw.db.Atomic(func(_ context.Context, tx *sql.Tx) error {
+		return persistBuilder(tx, rnd, &buildr)
+	})
+
+	// We ignore persisting errors because we still want to try and use our successfully generated builder,
+	// even if, for some reason, persisting it failed.
+	if err != nil {
+		spw.log.Errorf("loadOrCreateBuilder(%d): failed to insert builder into database: %v", rnd, err)
+	}
+
+	return buildr, nil
+}
+
+func (spw *Worker) loadSignaturesIntoBuilder(buildr *builder) error {
+	var sigs []pendingSig
+	err := spw.db.Atomic(func(ctx context.Context, tx *sql.Tx) error {
+		var err2 error
+		sigs, err2 = getPendingSigsForRound(tx, basics.Round(buildr.Round))
+		return err2
+	})
+	if err != nil {
+		return err
+	}
+
+	for _, sig := range sigs {
+		err = buildr.insertSig(&sig, false)
+		if err != nil {
+			spw.log.Warn(err)
+		}
+	}
+	return nil
+}
+
+func createBuilder(rnd basics.Round, votersFetcher ledgercore.LedgerForSPBuilder) (builder, error) {
```

Comment:
`createBuilder` is always followed by `persistBuilder`.
It is appropriate to more `persistBuilder` inside `createBuilder`.

Reply:
this is was like this before. The point is only `persistBuilder` part of the old function requires DB access and it is nicely used in tests